### PR TITLE
COMPASS-4345: Add document findOneAndUpdate helpers

### PR DESCRIPTION
**Pull request by Anemy**
COMPASS-4345

This PR adds a few helpers to hadron-document to help it be used in `compass-crud` with `findOneAndUpdate` commands. We're adding a helper that allow the developer to be able to more granularly understand the original document and the updated document. These helpers should help with the situation of checking whether a document was changed in the background while it was shown to a user. The new function to be used from `document` is `generateOriginal`.

Here's how they can be used to ensure we update the document we're showing the user and don't replace new fields that may have been added in the background:
```js
const originalDoc = doc.generateOriginal();
const newDoc = doc.generate();

const originalFieldsThatWillBeUpdated = getOriginalKeysAndValuesForFieldsThatWereUpdated(originalDoc, newDoc);
const setUpdateObject = generateSetUpdateObject(originalDoc, newDoc);
const unsetUpdateObject = generateUnsetUpdateObject(originalDoc, newDoc);
const updateObject = { };
if (setUpdateObject && Object.keys(setUpdateObject).length > 0) {
  updateObject.$set = setUpdateObject;
}
if (unsetUpdateObject && Object.keys(unsetUpdateObject).length > 0) {
  updateObject.$unset = unsetUpdateObject;
}

const opts = { returnOriginal: false, promoteValues: false };

this.dataService.findOneAndUpdate(
  'database.collection', // namespace
  { // query (search for the original document to be sure we're not overwriting new data)
    _id: documentId,
    ...originalFieldsThatWillBeUpdated
  },
  updateObject, // Contains the $set and $unset for changes on the document.
  opts,
  (error, d) => {
    if (error) {
      // Error occurred (connection failed or user doesn't have permission etc.).
    } else if (d) {
      // Update succeeded, d is the new document.
    } else {
      // Update failed to return new document, document was changed or no longer exists.
    }
  }
);

```

Next PR will be in [compass-crud](https://github.com/mongodb-js/compass-crud) using these helpers. It will also have the ability for an override for when the update is blocked because of document changes in the background. (It will perform the `findOneAndUpdate` without the original keys and values in the query).

Here's the work in progress on `compass-crud` where this is used: https://github.com/mongodb-js/compass-crud/compare/COMPASS-4345/use-find-one-and-update-when-updating-docs?expand=1#diff-61a4a14487b5a3b85978435e1dfb548eR341 (a bit outdated)

### CODE REVIEW

File: lib/object-generator.js [hunk 1]
```diff
@@ -80,6 +81,194 @@ var ObjectGenerator = function () {
...
+            var element = _step3.value;
+
+            if (!element.isRemoved() && element.currentKey !== '' && element.isModified()) {
+              // Include the full modified element. This way if a nested field
+              // has been altered (changed/added/removed) it is set at the top
+              // most level and exists in the set update document.
+              object[element.currentKey] = element.generateObject();
+            }
+          }
+        } catch (err) {
```
Comment (mcasimir) [inline lib/object-generator.js:134]: this is quite a bit of duplication, I think we can refactor it in a function that has a good name

File: lib/object-generator.js [hunk 2]
```diff
@@ -80,6 +81,194 @@ var ObjectGenerator = function () {
           for (var _iterator2 = elements[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
             var element = _step2.value;
 
+            object[element.key] = element.generateOriginalObject();
+          }
+        } catch (err) {
+          _didIteratorError2 = true;
```
Comment (mcasimir) [inline lib/object-generator.js:87]: Why don't we switch to `let` and `const` here and remove the `_` and numbers in variables?

File: lib/object-generator.js (hunk 1 above)
Comment (Anemy) [reply lib/object-generator.js:134]: Ah I think that's the `lib` file which is generated - they're checked into the repo. Here's the same method from `src`: https://github.com/mongodb-js/hadron-document/pull/71/files/adb6f357bfba7fe094d9d273751051662387680d#diff-c5b07fee6b95360111f7aa50295e7229R36

**Comment by Anemy** [comment]
Chatted with @mcasimir - going to remove most of this code except for the generateOriginal and then do diff in compass-crud to build the query there. (Also chatted about edge cases like when a document is removed vs edited in the background, how the set will work (and replacing if the doc doesn't exist)).

**Comment by Anemy** [comment]
@mcasimir Removed the set/unset helpers.

**Pull request merged**